### PR TITLE
add panorama pipeline and tiled VAE for ORT

### PR DESCRIPTION
**Pull request by ssube**
# What does this PR do?

Adds a new SD v1/v2 panorama pipeline based on the multi-diffusion paper and a tiled VAE for ORT, like the SDXL pipeline we've been discussing in https://github.com/huggingface/optimum/pull/1364.

The pipeline is based on https://multidiffusion.github.io/ and https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py, and supports txt2img, img2img, and inpainting.

```python
from optimum.onnxruntime import ORTStableDiffusionPanoramaPipeline

model_id = "runwayml/stable-diffusion-v1-5"
pipeline = ORTStableDiffusionPanoramaPipeline.from_pretrained(model_id)

# patch pipeline with tiled VAE
pipeline.vae_decoder = ORTModelTiledVaeWrapper(pipeline.vae_decoder, True, 64, 0.5)
pipeline.vae_encoder = ORTModelTiledVaeWrapper(pipeline.vae_encoder, False, 64, 0.5)

# run txt2img
prompt = "sailing ship in storm by Leonardo da Vinci"
image = pipeline(prompt).images[0]

# you can use the same cached pipeline for img2img
image = pipeline(prompt, image=image, strength=0.5).images[0]
```

The reference pipeline, https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py, does not offer img2img or inpainting, so we might have to mark those as experimental features. :see_no_evil: I find that panoramic inpainting works very well for outpainting large areas.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

You can run the new tests with `python -m pytest tests/onnxruntime/test_stable_diffusion_pipeline.py -k 'ORTStableDiffusionPanoramaPipelineTest' -s`. The tests depend on the DDIM scheduler, PNDM does not like re-running the same step repeatedly. 

I still need TODO:

- [ ] reduce difference with reference pipeline (I think this has to do with the window size)
- [ ] add examples for img2img and inpaint
- [ ] add tests for the tiled VAE

### CODE REVIEW

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_panorama.py [hunk 1]
```diff
@@ -0,0 +1,1099 @@
...
+
+        if output_type == "pil":
+            image = self.numpy_to_pil(image)
+
+        if not return_dict:
+            return (image, has_nsfw_concept)
+
+        return StableDiffusionPipelineOutput(images=image, nsfw_content_detected=has_nsfw_concept)
+
+    def __call__(
```
Comment (echarlaix) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion_panorama.py:1081]: I think we can keep the official text2img pipeline here and create two other pipelines (inheriting from `StableDiffusionPanoramaPipelineMixin`) specific to img2img and inpaint with each having their own `__call__`  method (+ would make sense to have them in the examples directly like done in https://github.com/huggingface/diffusers/tree/main/examples/community) at least until we don't have tests to verify everything is working as expected

File: tests/onnxruntime/test_stable_diffusion_pipeline.py [hunk 2]
```diff
@@ -483,3 +486,75 @@ def test_vae_image_processor_pil(self):
...
+        kwargs = {
+            "prompt": "sailing ship in storm by Leonardo da Vinci",
+            "num_inference_steps": 1,
+            "num_images_per_prompt": num_images_per_prompt,
+            "height": height,
+            "width": width,
+            "guidance_scale": 0.1,
+        }
+
+        for output_type in ["np"]:
```
Comment (echarlaix) [inline tests/onnxruntime/test_stable_diffusion_pipeline.py:533]: Thanks for the tests, they look great ! Could make sense to also add :  
```suggestion
        for output_type in ["np", "latent"]:
```

**Review by echarlaix — COMMENTED** [review]
Very cool, thanks a lot @ssube

File: docs/source/onnxruntime/usage_guides/models.mdx [hunk 3]
```diff
@@ -185,6 +185,24 @@ prompt = "Face of a yellow cat, high resolution, sitting on a park bench"
...
+```python
+import torch
+from optimum.onnxruntime import ORTStableDiffusionPanoramaPipeline, ORTModelTiledVaeWrapper
+
+model_id = "runwayml/stable-diffusion-v1-5"
+pipeline = ORTStableDiffusionPanoramaPipeline.from_pretrained(model_id, revision="onnx")
+
+# patch pipeline with tiled VAE
+pipeline.vae_decoder = ORTModelTiledVaeWrapper(pipeline.vae_decoder, True, 64, 0.5)
+pipeline.vae_encoder = ORTModelTiledVaeWrapper(pipeline.vae_encoder, False, 64, 0.5)
```
Comment (echarlaix) [inline docs/source/onnxruntime/usage_guides/models.mdx:199]: also I think we can integrate this in [`DiffusionPipelineMixin`](https://github.com/huggingface/optimum/blob/v1.13.2/optimum/pipelines/diffusers/pipeline_utils.py#L29) directly so that we can have instead : 

```suggestion
pipeline.enable_vae_tiling(64, 0.5)
```

Comment (echarlaix) [reply docs/source/onnxruntime/usage_guides/models.mdx:199]: let me know if you need any support on that
